File: kotlin/0374-guess-number-higher-or-lower.kt
/** 
 * The API guess is defined in the parent class.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * fun guess(num:Int):Int {}
 */

class Solution:GuessGame() {
    override fun guessNumber(n:Int):Int {
        var l = 1
        var r = n
        while ( true ) {
            val mid = l + (r - l) / 2
            val res = guess(mid)
            if(res == 0) return mid
            if(res > 0) l = mid + 1 
            if(res < 0) r = mid - 1
        }
    }
}

File: rust/0374-guess-number-higher-or-lower.rs
/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * unsafe fn guess(num: i32) -> i32 {}
 */

impl Solution {
    unsafe fn guessNumber(n: i32) -> i32 {
        Self::binary_search(1, n)
    }

    unsafe fn binary_search(left: i32, right: i32) -> i32 {
        let mid = left + (right - left ) / 2;

        if guess(mid) < 0 {
            Self::binary_search(1, mid)
        } else if guess(mid) > 0 {
            Self::binary_search(mid + 1, right)
        } else {
            mid
        }
    }
}
